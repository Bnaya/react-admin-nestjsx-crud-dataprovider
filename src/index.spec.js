import createDataProvider, { encodeParamsInResource } from "./index";

describe("Basic tests", () => {
  test("Basic test 1", async () => {
    // @ts-ignore
    const httpClient = jest.fn(async (type, resource, params) => {
      return {
        json: {
          id: 1
        },
        ...params
      };
    });

    const dataProvider = createDataProvider(
      "https://test.test/api",
      httpClient
    );

    expect(
      await dataProvider("GET_ONE", "books", { id: "5" })
    ).toMatchSnapshot();
    expect(await asyncResultsToValue(httpClient)).toMatchSnapshot();
  });
});

describe("encodeParamsInResource tests", () => {
  test("encodeParamsInResource test 1", async () => {
    expect(
      encodeParamsInResource("books", {
        fields: ["id", "name", "year"],
        join: [
          {
            field: "pages",
            select: ["number", "words"]
          },
          {
            field: "author",
            select: ["id", "name"]
          },
          {
            field: "author.favoriteFood",
            select: ["id", "name"]
          }
        ]
      })
    ).toMatchSnapshot();
  });
});

describe("encoded params tests", () => {
  test("encoded params test 1", async () => {
    // @ts-ignore
    const httpClient = jest.fn(async (type, resource, params) => {
      return {
        json: {
          id: 1
        },
        ...params
      };
    });

    const dataProvider = createDataProvider(
      "https://test.test/api",
      httpClient
    );

    const resourceWithParams = encodeParamsInResource("books", {
      fields: ["id", "name", "year"],
      join: [
        {
          field: "pages",
          select: ["number", "words"]
        },
        {
          field: "author",
          select: ["id", "name"]
        },
        {
          field: "author.favoriteFood",
          select: ["id", "name"]
        }
      ]
    });

    expect(
      await dataProvider("GET_ONE", resourceWithParams, { id: "5" })
    ).toMatchSnapshot();
    expect(await asyncResultsToValue(httpClient)).toMatchSnapshot();
  });
});

// @ts-ignore
async function asyncResultsToValue(jestFned) {
  // @ts-ignore
  jestFned.mock.results = await Promise.all(
    (jestFned.mock.results || []).map(
      /**
       * @param {{ value: any; type: any; }} result
       */ async result => {
        if (result.value instanceof Promise) {
          return {
            type: result.type,
            wasPromise: true,
            value: await result.value
          };
        }

        return result;
      }
    )
  );

  return jestFned;
}
