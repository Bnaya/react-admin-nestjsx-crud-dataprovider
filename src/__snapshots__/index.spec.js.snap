// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic tests Basic test 1 1`] = `
Object {
  "data": Object {
    "id": 1,
  },
}
`;

exports[`Basic tests Basic test 1 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://test.test/api/books/5?",
      Object {},
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "json": Object {
          "id": 1,
        },
      },
      "wasPromise": true,
    },
  ],
}
`;

exports[`encodeParamsInResource tests encodeParamsInResource test 1 1`] = `"books_._._._{\\"fields\\":[\\"id\\",\\"name\\",\\"year\\"],\\"join\\":[{\\"field\\":\\"pages\\",\\"select\\":[\\"number\\",\\"words\\"]},{\\"field\\":\\"author\\",\\"select\\":[\\"id\\",\\"name\\"]},{\\"field\\":\\"author.favoriteFood\\",\\"select\\":[\\"id\\",\\"name\\"]}]}"`;

exports[`encoded params tests encoded params test 1 1`] = `
Object {
  "data": Object {
    "id": 1,
  },
}
`;

exports[`encoded params tests encoded params test 1 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://test.test/api/books/5?join[]=pages||number,words&join[]=author||id,name&join[]=author.favoriteFood||id,name&fields=id,name,year",
      Object {},
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "json": Object {
          "id": 1,
        },
      },
      "wasPromise": true,
    },
  ],
}
`;
